### PR TITLE
blocker: hot fix keplr sign failed

### PR DESCRIPTION
**Pull request by vuong177**
Fix failed sign in kelpr wallet.

### CODE REVIEW

**Review by faddat — APPROVED** [review]

**Review by nghuyenthevinh2000 — APPROVED** [review]

**Comment by faddat** [comment]
Please test this on a local testnet, we should make sure that fixing this didn't break the account model changes, but I don't think it does.

**Comment by vuong177** [comment]
I've tested

**Comment by vuong177** [comment]
I asked Son for fix in their multisig app, double check

File: app/ante.go
Comment (mustafapylons) [inline app/ante.go:40]: can we remove dead code here?

**Review by mustafapylons — COMMENTED** [review]
one change necessary and this can be approved after that.

File: app/ante.go
Comment (faddat) [reply app/ante.go:40]: I would prefer not to actually I need to review something, good catch

**Comment by faddat** [comment]
have a look at this file:


```go
package simapp

import (
	sdk "github.com/cosmos/cosmos-sdk/types"
	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
	"github.com/cosmos/cosmos-sdk/x/auth/ante"

	ibcante "github.com/cosmos/ibc-go/v5/modules/core/ante"
	"github.com/cosmos/ibc-go/v5/modules/core/keeper"
)

// HandlerOptions extend the SDK's AnteHandler options by requiring the IBC keeper.
type HandlerOptions struct {
	ante.HandlerOptions

	IBCKeeper *keeper.Keeper
}

// NewAnteHandler creates a new ante handler
func NewAnteHandler(options HandlerOptions) (sdk.AnteHandler, error) {
	if options.AccountKeeper == nil {
		return nil, sdkerrors.Wrap(sdkerrors.ErrLogic, "account keeper is required for AnteHandler")
	}
	if options.BankKeeper == nil {
		return nil, sdkerrors.Wrap(sdkerrors.ErrLogic, "bank keeper is required for AnteHandler")
	}
	if options.SignModeHandler == nil {
		return nil, sdkerrors.Wrap(sdkerrors.ErrLogic, "sign mode handler is required for AnteHandler")
	}

	anteDecorators := []sdk.AnteDecorator{
		ante.NewSetUpContextDecorator(), // outermost AnteDecorator. SetUpContext must be called first
		ante.NewExtensionOptionsDecorator(options.ExtensionOptionChecker),
		ante.NewValidateBasicDecorator(),
		ante.NewTxTimeoutHeightDecorator(),
		ante.NewValidateMemoDecorator(options.AccountKeeper),
		ante.NewConsumeGasForTxSizeDecorator(options.AccountKeeper),
		ante.NewDeductFeeDecorator(options.AccountKeeper, options.BankKeeper, options.FeegrantKeeper, options.TxFeeChecker),
		ante.NewSetPubKeyDecorator(options.AccountKeeper), // SetPubKeyDecorator must be called before all signature verification decorators
		ante.NewValidateSigCountDecorator(options.AccountKeeper),
		ante.NewSigGasConsumeDecorator(options.AccountKeeper, options.SigGasConsumer),
		ante.NewSigVerificationDecorator(options.AccountKeeper, options.SignModeHandler),
		ante.NewIncrementSequenceDecorator(options.AccountKeeper),
		ibcante.NewRedundantRelayDecorator(options.IBCKeeper),
	}

	return sdk.ChainAnteDecorators(anteDecorators...), nil
}
```

it is from the ibc-go simapp.  I think that we need to follow it.

File: app/ante.go
Comment (faddat) [reply app/ante.go:40]: I've made some changes here and will pass to our team to review.  This file was pretty badly out of spec, and the last time I'd touched it was the 46 upgrade.  Mustafa, this was a really, really good catch.

**Review by faddat — CHANGES_REQUESTED** [review]
Hey guys so for the time being I have blocked the merge of this one. The reason is that there was some pretty rough legacy stuff in the anti-handler. So what I have done is I got the work started on a full re-factor of the anti-handler and all of the things that call it from app.go and it's actually a pretty significant piece of work that this is going to need I don't know exactly how long it is going to take.

* I found an issue with the store keys, look in app.go you will find that there are some store keys missing, please fill those in
* I couldn't find pylons in the usual keeper declaration list so I added it this really didn't make a lot of sense to me, if it wasn't there I'm not really sure how it didn't work before. it shouldn't have worked.
* the fix for the multi signature code is accurate and proper I think that doesn't need further refactoring and please have a look at my changes in the anti-handler to understand that better.
* somehow none of these were flagged as errors before even by the automated code analysis tools even past our reviews and so I have kind of a dark gas which is that there might be codes somewhere that we missed somehow that caused it to pass but that code that caused it to pass probably shouldn't exist so what I would like for the Notional team to do is to take this PR, look at it when you wake up today and make changes to it that you think address the issues that I've mentioned also I think you might find other issues because when I reach factored the anti-handler I've started to see issues pop-up left and right
* when this is finished we need to schedule a call with RNS to do a code reading session kind of like how Dev does for osmosis

**Comment by nghuyenthevinh2000** [comment]
> @nghuyenthevinh2000 I've gotta revert your commit here-- have a look at the changes that I made, and at the ibc-go antehandler code. The whole antehandler needs a refactor as do signficant sections of app.go

have a look at #419, without CustomSigVerificationDecorator, #419 may not work as intended. It seems to be a core logic for pylons.

**Comment by faddat** [comment]
yep, so add it to the antehandler stuff I put in there.

File: app/app.go
Comment (faddat) [inline app/app.go:258]: this

File: app/app.go
Comment (faddat) [inline app/app.go:268]: this

File: app/app.go
Comment (faddat) [inline app/app.go:267]: this

File: app/app.go
Comment (faddat) [inline app/app.go:391]: we need this for full blown ibc v5, but it worked without it.  that is bad and it is also mysterious.

File: app/app.go
Comment (faddat) [inline app/app.go:665]: this is the canonical way of passing to the antehandler, we should use it

File: app/app.go
Comment (faddat) [inline app/app.go:19]: we need this to pass to the antehandler in a canonical manner

**Comment by faddat** [comment]
The fix exposed so many issues.

**Comment by nghuyenthevinh2000** [comment]
yep, I totally agree with items you mention that need refactoring.

preferably, I want to separate the refactoring from the original hot fix of Vuong. It will block Son's multisig --> block the creation of 8 founder multisig address for bedrock (which can take lots of time) --> block bedrock clean.

by separating the refactoring from the original hot fix of Vuong, code development works will achieve higher parallel, thus more efficient.

**Comment by faddat** [comment]
Ok, I will unrevert, good call.  Please take the commentary here about the chains crazy behavior, and make issues and PR's <3.

**Comment by faddat** [comment]
Please begin work on the new pr from commit: 14968a7a5b7f48b96e9aaa6618765b531fb73cfe

File: app/ante.go
Comment (faddat) [reply app/ante.go]: short answer: for now, no.  I want that to stay exactly as it is and then we will refactor.

**Review by faddat — APPROVED** [review]

**Pull request merged**
